File: Leetcode/C++/22. Generate Parentheses/solution.cpp
// https://leetcode.com/problems/generate-parentheses/description/

#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<string> result;
    int length;
    vector<string> generateParenthesis(int n)
    {
        length = n;
        string p = "";
        search(0, 0, p);
        return result;
    }
    // Search like DFS
    void search(int left, int right, string &parenthesis)
    {
        if (left == right && left == length)
        {
            result.push_back(parenthesis);
        }
        if (left < length)
        {
            parenthesis += "(";
            search(left + 1, right, parenthesis);
            parenthesis.pop_back();
        }
        if (right < left)
        {
            parenthesis += ")";
            search(left, right + 1, parenthesis);
            parenthesis.pop_back();
        }
    }
};

File: Leetcode/Python/22. Generate Parentheses/solution.py
# https://leetcode.com/problems/generate-parentheses/


from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        def generate(left: int, right: int, parenthesis: str):
            if left == right == n:  # Add result complete parenthesis
                result.append(parenthesis)
                return
            if left < n:    # Add left parenthesis
                generate(left + 1, right, parenthesis + "(")
            # If number of right parenthesis > left parenthesis is invaild parenthesis so add right parenthesis when right < left
            if right < left:    
                generate(left, right + 1, parenthesis + ")")
        generate(0, 0, "")
        return result

